code/modules/admin/IsBanned.dm

```diff
@@ -0,0 +1,233 @@
+//Blocks an attempt to connect before even creating our client datum thing.
+
+//How many new ckey matches before we revert the stickyban to it's roundstart state
+//These are exclusive, so once it goes over one of these numbers, it reverts the ban
+#define STICKYBAN_MAX_MATCHES 15
+#define STICKYBAN_MAX_EXISTING_USER_MATCHES 3 //ie, users who were connected before the ban triggered
+#define STICKYBAN_MAX_ADMIN_MATCHES 1
+
+/world/IsBanned(key, address, computer_id, type, real_bans_only=FALSE)
+	if (!key || (!real_bans_only && (!address || !computer_id)))
+		if(real_bans_only)
+			return FALSE
+		log_access("Failed Login (invalid data): [key] [address]-[computer_id]")
+		return list("reason"="invalid login data", "desc"="Error: Could not check ban status, Please try again. Error message: Your computer provided invalid or blank information to the server on connection (byond username, IP, and Computer ID.) Provided information for reference: Username:'[key]' IP:'[address]' Computer ID:'[computer_id]'. (If you continue to get this error, please restart byond or contact byond support.)")
+	
+	if (type == "world")
+		return ..() //shunt world topic banchecks to purely to byond's internal ban system
+		
+	var/admin = FALSE
+	var/ckey = ckey(key)
+	
+	var/client/C = GLOB.directory[ckey]
+	if (C && ckey == C.ckey && computer_id == C.computer_id && address == C.address)
+		return //don't recheck connected clients.
+		
+	//IsBanned can get re-called on a user in certain situations, this prevents that leading to repeated messages to admins.
+	var/static/list/checkedckeys = list()
+	//magic voodo to check for a key in a list while also adding that key to the list without having to do two associated lookups
+	var/message = !checkedckeys[ckey]++
+
+	if(GLOB.admin_datums[ckey] || GLOB.deadmins[ckey])
+		admin = TRUE
+
+
+	//Whitelist
+	if(!real_bans_only && !C && CONFIG_GET(flag/usewhitelist))
+		if(!check_whitelist(ckey))
+			if (admin)
+				log_admin("The admin [key] has been allowed to bypass the whitelist")
+				if (message)
+					message_admins("<span class='adminnotice'>The admin [key] has been allowed to bypass the whitelist</span>")
+					addclientmessage(ckey,"<span class='adminnotice'>You have been allowed to bypass the whitelist</span>")
+			else
+				log_access("Failed Login: [key] - Not on whitelist")
+				return list("reason"="whitelist", "desc" = "\nReason: You are not on the white list for this server")
+
+	//Guest Checking
+	if(!real_bans_only && !C && IsGuestKey(key))
+		if (CONFIG_GET(flag/guest_ban))
+			log_access("Failed Login: [key] - Guests not allowed")
+			return list("reason"="guest", "desc"="\nReason: Guests not allowed. Please sign in with a byond account.")
+		if (CONFIG_GET(flag/panic_bunker) && SSdbcore.Connect())
```

Comment:
Your connecting to the database here before doing sql_enabled check.
Is that okay?

Reply:
It's not ideal, there's quite a bit I'd clean up on this proc, but this is /tg/ code and I prefer compatibility over clarity in this case. So when they improve stuff we'll have less work porting. This is not code we'll be accessing often anwyay.
I guess it means that the `flag/panic_bunker` assumes `flag/sql_enabled` to be true.